way-cooler-bg/src/main.rs

```diff
@@ -0,0 +1,285 @@
+use wayland_client::wayland::seat::{WlSeat, WlPointerEvent};
+use wayland_client::cursor::load_theme;
+use wayland_client::{EventIterator, Proxy};
+
+use byteorder::{NativeEndian, WriteBytesExt};
+
+wayland_env!(WaylandEnv,
+             compositor: WlCompositor,
+             shell: WlShell,
+             shm: WlShm,
+             seat: WlSeat
+);
+
+type BufferResult = Result<WlBuffer, ()>;
+
+/// Hack to deal with edge case between the two cursor buffer types...
+enum CursorBuffer {
+    /// The buffer has been leaked into memory.
+    /// This should eventually be fixed, but since it's just the cursor
+    /// it's not a big issue.
+    Null,
+    /// The buffer for the cursor, this can be destroyed so it should last
+    /// as long as the program (or until you replace it with another).
+    Buf(WlBuffer)
+}
+
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq)]
+/// Holds the bytes to represent a colored background.
+/// To be written into a wayland surface.
+pub struct Color(pub [u8; 4]);
+
+impl Color {
+    /// Generate a new color out of a u32.
+    /// E.G: 0xFFFFFF
+    pub fn from_u32(color: u32) -> Self {
+        unsafe { Color(transmute(color)) }
+    }
+
+    pub fn as_u32(&self) -> u32 {
+        unsafe { transmute(self.0)}
+    }
+}
+
+fn main() {
+    let args: Vec<_> = env::args().collect();
+    if args.len() < 3 {
+        println!("Please supply either a file path or a color (written in hex)");
+        println!("Please also supply a path to a cursor image");
+        exit(1);
+    }
+    let input = &args[1];
+    let cursor_path = &args[2];
+
+    let (display, iter) = get_display()
+        .expect("Unable to connect to a wayland compositor");
+    let (env, evt_iter) = WaylandEnv::init(display, iter);
+    let compositor = env.compositor.as_ref().map(|o| &o.0).unwrap();
+    let mut background_surface = compositor.create_surface();
+
+    // We need to hold on to this buffer, this holds the background image!
+    let shell = env.shell.as_ref().map(|o| &o.0).unwrap();
+    let shell_surface = shell.get_shell_surface(&background_surface);
+    shell_surface.set_class("Background".into());
+    // TODO Actually give it the path or something idk
+    shell_surface.set_title(input.clone());
+    let _background_buffer = if let Ok(color) = input.parse::<u32>() {
+        let color = Color::from_u32(color);
+        generate_solid_background(color, &mut background_surface, &env)
+    } else {
+        generate_image_background(input.as_str(), &mut background_surface, &env)
+    }.expect("could not generate image");
+
+    background_surface.commit();
+    background_surface.set_buffer_scale(1);
+    let mut cursor_surface = compositor.create_surface();
+    let _cursor_buffer = self::cursor_surface(cursor_path.as_str(), &mut cursor_surface, &env);
+    main_background_loop(background_surface, cursor_surface, evt_iter, &env);
+}
+
+fn weird_math(num: u8, third_num: u32) -> u8 {
```

Comment:
Yes, probably a bad name. I wasn't sure what that conversion process was called (it's the same one eg sway uses), but all I know is it goes from RGBA ->Some slightly different values for RGBA that are somehow correct :tm: